### DISCUSSION TITLE
How to construct a dataset using a fixed graph with varying node features samples?.

### DISCUSSION THREAD

**Discussion by t-smurf** [pyg-team/pytorch_geometric#5413] [Q&A]
Hello, 

I am trying to solve a regression problem using GNN. 
I have a dataset. I managed to model the problem as a static graph, with a fixed number of nodes, and fixed number of features. However, dataset contains  different size of samples for each node's features. 

To give an example, the following csv file contains 3 nodes {0, 1, 2}. However, we have 3 samples for node 1, and 2 samples for node 0 and 1 for node 2.

The size of this csv file is S x  F (Samples x Features)

The content of `dataset.csv`
```
node,feature1,feature2,feature3
0, 1, 2, 3
0, 1, 1, 3
1, 6, 5, 7
1, 0, 68, 1
1, 7, 9, 76
2, 7, 8, 9
```

I am trying to explore how to do that using PG. I checked this [discussion](https://github.com/pyg-team/pytorch_geometric/issues/817). My understanding is that we can batch node features (feature1,feature2,feature3 in my case) with shape [B, N, F] and a sparse single adjacency matrix edge_index with shape [2, E], 

```
edge_index = torch.tensor([[0, 1, 1, 2], [1, 0, 2, 1]])
x_row_1 =   torch.tensor([1, 2, 3])           # this is row 1 from dataset.csv presented above
x_row_2 =  torch.tensor([1, 1, 3])           # this is row 2
x_row_3 =  torch.tensor([ 6, 5, 7])         # this is row 3
x_row_4 =  torch.tensor([ 0, 68, 1])       # this is row 4
x_row_5 =  torch.tensor([7, 9, 76])        # this is row 5
x_row_6 =  torch.tensor([7, 8, 9])          # this is row 6

# loading each row that contains feature sample 
data_1  = Data(edge_index=edge_index, x=x_row_1)
data_2 = Data(edge_index=edge_index, x=x_row_2)
data_3 = Data(edge_index=edge_index, x=x_row_3)
data_4 = Data(edge_index=edge_index, x=x_row_4)
data_5 = Data(edge_index=edge_index, x=x_row_5)
data_6 = Data(edge_index=edge_index, x=x_row_6)

# batch them
batch = Batch.from_data_list([data_1, data_2, data_3, data_4, data_5, data_6])
```
We can use InMemoryDataset class to wrap this, but I just want to give a quick example. 

*Question:*
In case we have different size of samples for each node's features, is this OK, or does the implementation consider the number of samples to be equal when batching them?.

**Comment by rusty1s** [comment]
Sorry for the late reply. I assume you want to combine all different samples across all nodes, right? In this case, you should do
```python
data1 = Data(edge_index=edge_index, x=torch.stack([x_row_1, x_row_3, x_row_6], dim=0))
data1 = Data(edge_index=edge_index, x=torch.stack([x_row_2, x_row_3, x_row_6], dim=0))
data1 = Data(edge_index=edge_index, x=torch.stack([x_row_1, x_row_4, x_row_6], dim=0))
data1 = Data(edge_index=edge_index, x=torch.stack([x_row_2, x_row_4, x_row_6], dim=0))
data1 = Data(edge_index=edge_index, x=torch.stack([x_row_1, x_row_5, x_row_6], dim=0))
data1 = Data(edge_index=edge_index, x=torch.stack([x_row_2, x_row_5, x_row_6], dim=0))
```

**Reply by t-smurf** [reply]
I am not sure if the question was clear. I will give it another try.

In the `dataset.csv` presented above, we have 3 nodes, 3 features, and 9 samples. Number of samples for each node is not equal.

Number of samples of node 0 is 2. Features (feature1,feature2,feature3)  for node 0 are: 
```
node,feature1,feature2,feature3
0, 1, 2, 3
0, 1, 1, 3
```


Number of samples of node 1 is 3. Features (feature1,feature2,feature3)  for node 1 are: 
```
node,feature1,feature2,feature3
1, 6, 5, 7
1, 0, 68, 1
1, 7, 9, 76
```

Number of samples of node 2 is 3. Features (feature1,feature2,feature3)  for node 12are: 
```
node,feature1,feature2,feature3
2, 7, 8, 9 
```


Then, if we represent each row as a tensor, we can get something similar to the following 
```
x_row_1 =   torch.tensor([1, 2, 3])           # this is row 1 from dataset.csv presented above
x_row_2 =  torch.tensor([1, 1, 3])           # this is row 2
x_row_3 =  torch.tensor([ 6, 5, 7])         # this is row 3
x_row_4 =  torch.tensor([ 0, 68, 1])       # this is row 4
x_row_5 =  torch.tensor([7, 9, 76])        # this is row 5
x_row_6 =  torch.tensor([7, 8, 9])          # this is row 6
```
where 
```
node 0 can be represented by x_row_1, x_row_2
node 1 can be represented by x_row_3, x_row_4, x_row_5
node 2 can be represented by x_row_6

```

what you are saying is that we need to find all possible combinations  and batch them.

**Reply by Padarn** [reply]
Thats my understanding of @rusty1s 's suggestion: But I'm curious why you have uneven samples from the nodes? It hard to understand if its a good idea or not.

**Reply by t-smurf** [reply]
@Padarn Thanks. 
I am trying to solve a problem using (1) MLP (2) GNN. The task is a regression task, where the data has been collected randomly for each node.
For regression using MLP, this is common to have an uneven samples.

**Reply by Padarn** [reply]
I see. I think if every combination of features is a valid training example, then what is suggested already is your best bet.
